pandas/core/interval.py

```diff
@@ -0,0 +1,521 @@
+        """
+        # TODO: exclude periods?
+        if self.closed not in _VALID_CLOSED:
+            raise ValueError("invalid options for 'closed': %s" % self.closed)
+        if len(self.left) != len(self.right):
+            raise ValueError('left and right must have the same length')
+        left_valid = notnull(self.left)
+        right_valid = notnull(self.right)
+        if not (left_valid == right_valid).all():
+            raise ValueError('missing values must be missing in the same '
+                             'location both left and right sides')
+        if not (self.left[left_valid] <= self.right[left_valid]).all():
+            raise ValueError('left side of interval must be <= right side')
+
+    def _simple_new(cls, values, name=None, **kwargs):
+        # ensure we don't end up here (this is a superclass method)
+        raise NotImplementedError
+
+    def _cleanup(self):
+        pass
+
+    @property
+    def _engine(self):
+        raise NotImplementedError
+
+    @cache_readonly
+    def _tree(self):
+        return IntervalTree(self.left, self.right, closed=self.closed)
+
+    @property
+    def _constructor(self):
+        return type(self).from_intervals
+
+    @classmethod
+    def from_breaks(cls, breaks, closed='right', name=None):
+        """
+        Construct an IntervalIndex from an array of splits
+
+        Parameters
+        ----------
+        breaks : array-like (1-dimensional)
+            Left and right bounds for each interval.
+        closed : {'left', 'right', 'both', 'neither'}, optional
+            Whether the intervals are closed on the left-side, right-side, both
+            or neither. Defaults to 'right'.
+        name : object, optional
+            Name to be stored in the index.
+
+        Examples
+        --------
+
+        >>> IntervalIndex.from_breaks([0, 1, 2, 3])
+        IntervalIndex(left=[0, 1, 2],
+                      right=[1, 2, 3],
+                      closed='right')
+        """
+        return cls(breaks[:-1], breaks[1:], closed, name)
+
+    @classmethod
+    def from_intervals(cls, data, name=None):
+        """
+        Construct an IntervalIndex from a 1d array of Interval objects
+
+        Parameters
+        ----------
+        data : array-like (1-dimensional)
+            Array of Interval objects. All intervals must be closed on the same
+            sides.
+        name : object, optional
+            Name to be stored in the index.
+
+        Examples
+        --------
+
+        >>> IntervalIndex.from_intervals([Interval(0, 1), Interval(1, 2)])
+        IntervalIndex(left=[0, 1],
+                      right=[1, 2],
+                      closed='right')
+
+        The generic Index constructor work identically when it infers an array
+        of all intervals:
+
+        >>> Index([Interval(0, 1), Interval(1, 2)])
+        IntervalIndex(left=[0, 1],
+                      right=[1, 2],
+                      closed='right')
```

Comment:
This is more of a reminder to myself.  Will want to have all of the intervals have the same closure.
This is already being raised in `intervals_to_interval_bounds`, so we'll want a `Raises` section in this doc-string